### PR TITLE
Switch to YYJSON 

### PR DESCRIPTION
**Pull request by AlexKlo**
### Pull request checklist

- [x] Did you bump the project version?
- [x] Did you add a description to the Pull Request?
- [x] Did you add new tests?
- [x] Did you add reviewers?

### Old deser
```
@benchmark deser_json(ExchangeInfoData, json_str)
BenchmarkTools.Trial: 7 samples with 1 evaluation.
 Range (min … max):  788.652 ms … 850.092 ms  ┊ GC (min … max): 27.89% … 31.28%
 Time  (median):     820.066 ms               ┊ GC (median):    30.80%
 Time  (mean ± σ):   822.370 ms ±  20.313 ms  ┊ GC (mean ± σ):  30.34% ±  1.43%

  █                  █         █ █            █   █           █  
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█▁▁▁▁▁▁▁▁▁█▁█▁▁▁▁▁▁▁▁▁▁▁▁█▁▁▁█▁▁▁▁▁▁▁▁▁▁▁█ ▁
  789 ms           Histogram: frequency by time          850 ms <

 Memory estimate: 667.81 MiB, allocs estimate: 4027315.
```

### New deser
```
@benchmark deser_json(ExchangeInfoData, json_str)
BenchmarkTools.Trial: 18 samples with 1 evaluation.
 Range (min … max):  276.343 ms … 308.855 ms  ┊ GC (min … max): 0.00% … 2.33%
 Time  (median):     288.023 ms               ┊ GC (median):    1.47%
 Time  (mean ± σ):   289.823 ms ±   9.065 ms  ┊ GC (mean ± σ):  1.39% ± 1.41%

  ▁    ▁ ▁ ▁    █▁▁  ▁    █     ▁▁         ▁  ▁ ▁▁            ▁  
  █▁▁▁▁█▁█▁█▁▁▁▁███▁▁█▁▁▁▁█▁▁▁▁▁██▁▁▁▁▁▁▁▁▁█▁▁█▁██▁▁▁▁▁▁▁▁▁▁▁▁█ ▁
  276 ms           Histogram: frequency by time          309 ms <

 Memory estimate: 31.92 MiB, allocs estimate: 394033.
```

```
julia> versioninfo()
Julia Version 1.11.1
Commit 8f5b7ca12ad (2024-10-16 10:53 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 20 × 12th Gen Intel(R) Core(TM) i7-12700H
  WORD_SIZE: 64
  LLVM: libLLVM-16.0.6 (ORCJIT, alderlake)
Threads: 1 default, 0 interactive, 1 GC (on 20 virtual cores)
```

### CODE REVIEW

File: src/De/DeJson.jl [hunk 1]
```diff
@@ -2,15 +2,326 @@ module DeJson
...
+    elseif yyjson_is_int(val_ptr)
+        yyjson_get_int(val_ptr)
+    elseif yyjson_is_bool(val_ptr)
+        yyjson_get_bool(val_ptr)
+    elseif yyjson_is_obj(val_ptr)
+        deser(Dict{String,Any}, val_ptr)
+    elseif yyjson_is_arr(val_ptr)
+        deser(Vector{Any}, val_ptr)
+    end
+end
```
Comment (gryumov) [inline src/De/DeJson.jl:29]:
```suggestion
function deser(::Type{Any}, value_ptr::Ptr{YYJSONVal})
    return if yyjson_is_str(value_ptr)
        unsafe_string(yyjson_get_str(value_ptr))
    elseif yyjson_is_raw(value_ptr)
        unsafe_string(yyjson_get_raw(value_ptr))
    elseif yyjson_is_real(value_ptr)
        yyjson_get_real(value_ptr)
    elseif yyjson_is_int(value_ptr)
        yyjson_get_int(value_ptr)
    elseif yyjson_is_bool(value_ptr)
        yyjson_get_bool(value_ptr)
    elseif yyjson_is_obj(value_ptr)
        deser(Dict{String,Any}, value_ptr)
    elseif yyjson_is_arr(value_ptr)
        deser(Vector{Any}, value_ptr)
    else
        throw(WrongType("Unsupported JSON type for `Any`."))
    end
end
```

File: src/De/DeJson.jl [hunk 2]
```diff
@@ -2,15 +2,326 @@ module DeJson
...
+    elseif yyjson_is_raw(val_ptr)
+        unsafe_string(yyjson_get_raw(val_ptr))
+    elseif yyjson_is_real(val_ptr)
+        repr(yyjson_get_num(val_ptr))
+    elseif yyjson_is_int(val_ptr)
+        repr(Int64(yyjson_get_num(val_ptr)))
+    elseif yyjson_is_bool(val_ptr)
+        repr(yyjson_get_bool(val_ptr))
+    end
+end
```
Comment (gryumov) [inline src/De/DeJson.jl:57]:
```suggestion
function deser(::PrimitiveType, ::Type{T}, value_ptr::Ptr{YYJSONVal}) where {T<:AbstractString}
    return if yyjson_is_str(value_ptr)
        unsafe_string(yyjson_get_str(value_ptr))
    elseif yyjson_is_raw(value_ptr)
        unsafe_string(yyjson_get_raw(value_ptr))
    else
        throw(WrongType("Expected a string for type `AbstractString`."))
    end
end
```

File: src/De/DeJson.jl [hunk 3]
```diff
@@ -2,15 +2,326 @@ module DeJson
...
+    elseif yyjson_is_raw(val_ptr)
+        tryparse(T, unsafe_string(yyjson_get_raw(val_ptr)))
+    elseif yyjson_is_real(val_ptr)
+        T(yyjson_get_num(val_ptr))
+    elseif yyjson_is_int(val_ptr)
+        T(Int64(yyjson_get_num(val_ptr)))
+    elseif yyjson_is_bool(val_ptr)
+        T(yyjson_get_bool(val_ptr))
+    end
+end
```
Comment (gryumov) [inline src/De/DeJson.jl:71]:
```suggestion
function deser(::PrimitiveType, ::Type{T}, value_ptr::Ptr{YYJSONVal}) where {T<:Number}
    return if yyjson_is_real(value_ptr) || yyjson_is_int(value_ptr)
        T(yyjson_get_num(value_ptr))
    else
        throw(WrongType("Expected a number for type `Number`."))
    end
end
```

File: src/De/DeJson.jl [hunk 4]
```diff
@@ -2,15 +2,326 @@ module DeJson
...
+
+function deser(::PrimitiveType, ::Type{T}, val_ptr::Ptr{YYJSONVal}) where {T<:Symbol}
+    return if yyjson_is_str(val_ptr)
+        Symbol(unsafe_string(yyjson_get_str(val_ptr)))
+    elseif yyjson_is_raw(val_ptr)
+        Symbol(unsafe_string(yyjson_get_raw(val_ptr)))
+    end
+end
```
Comment (gryumov) [inline src/De/DeJson.jl:97]:
```suggestion
function deser(::PrimitiveType, ::Type{T}, value_ptr::Ptr{YYJSONVal}) where {T<:Symbol}
    return if yyjson_is_str(value_ptr)
        Symbol(unsafe_string(yyjson_get_str(value_ptr)))
    elseif yyjson_is_raw(value_ptr)
        Symbol(unsafe_string(yyjson_get_raw(value_ptr)))
    else
        throw(WrongType("Expected a string for type `Symbol`."))
    end
end
```

File: src/De/DeJson.jl [hunk 5]
```diff
@@ -2,15 +2,326 @@ module DeJson
...
+
+# NullType
+
+function deser(::NullType, ::Type{Nothing}, val_ptr::Ptr{YYJSONVal})
+    return if yyjson_is_null(val_ptr)
+        nothing
+    end
+end
```
Comment (gryumov) [inline src/De/DeJson.jl:105]:
```suggestion
function deser(::NullType, ::Type{Nothing}, value_ptr::Ptr{YYJSONVal})
    return if yyjson_is_null(value_ptr)
        nothing
    else
        throw(WrongType("Expected null for type `Nothing`."))
    end
end
```

File: src/De/DeJson.jl [hunk 6]
```diff
@@ -2,15 +2,326 @@ module DeJson
...
+        end
+
+        type_elements[i] = if isnothing(val) || ismissing(val) || isempty(T, val)
+            nulltype(type)
+        else
+            val
+        end
+    end
+    return T(type_elements...)
+end
```
Comment (gryumov) [inline src/De/DeJson.jl:269]:
```suggestion
function deser_obj(::CustomType, ::Type{T}, obj_ptr::Ptr{YYJSONVal}) where {T}
    field_values = Vector{Any}(undef, fieldcount(T))
    for (index, field_type, field_name) in zip(eachindex(field_values), fieldtypes(T), fieldnames(T))
        key = custom_name(T, Val(field_name))
        value_ptr = yyjson_obj_get(obj_ptr, key)
        value = if value_ptr == C_NULL
            default_value(T, Val(field_name))
        else
            eldeser(T, field_type, key, value_ptr)
        end
        field_values[index] = if isnothing(value) || ismissing(value) || isempty(T, value)
            nulltype(field_type) 
        else
            value
        end
    end
    return T(field_values...)
end
```

File: src/De/DeJson.jl [hunk 7]
```diff
@@ -2,15 +2,326 @@ module DeJson
...
+        bitwise_read_flag(; kw...),
+        C_NULL,
+        pointer_from_objref(err),
+    )
+    if doc_ptr == C_NULL
+        yyjson_doc_free(doc_ptr)
+        throw(err)
+    end
+    return doc_ptr
+end
```
[the remark this answers is not in the document]
Comment (AlexKlo) [reply src/De/DeJson.jl:317]: We shouldn't

File: src/De/DeJson.jl (hunk 1 above)
Comment (AlexKlo) [reply src/De/DeJson.jl]: I didn't add an error throw here because it happens outside

File: src/De/DeJson.jl (hunk 2 above)
Comment (AlexKlo) [reply src/De/DeJson.jl]: WrongType expects a lot of arguments so decided to throw an julia `error`.

File: src/De/DeJson.jl (hunk 3 above)
Comment (AlexKlo) [reply src/De/DeJson.jl]: Same with WrongType

File: src/De/DeJson.jl (hunk 4 above)
Comment (AlexKlo) [reply src/De/DeJson.jl]: Same with WrongType

File: src/De/DeJson.jl (hunk 5 above)
Comment (AlexKlo) [reply src/De/DeJson.jl]: Same with WrongType

File: test/deser.jl [hunk 8]
```diff
@@ -659,7 +659,7 @@ using Test, Dates
         end
 
         exp_str = """ {"correlation_id":2,"method":"subscribe.status","payload":{}} """
-        @test_throws "WrongType: for 'Message{Nothing}' value 'Dict{String, Any}()' has wrong type 'payload::Dict{String, Any}', must be 'payload::Nothing'" Serde.deser_json(
+        @test_throws "WrongType: for 'Message{Nothing}' value 'Ptr{YYJSONVal}' has wrong type 'payload::Dict{String, Any}', must be 'payload::Nothing'" Serde.deser_json(
```
Comment (AlexKlo) [inline test/deser.jl:662]: @gryumov what do you think about such an error message?

File: src/De/DeJson.jl [hunk 9]
```diff
@@ -2,15 +2,359 @@ module DeJson
...
+        any(x -> x <: T, Base.uniontypes(E))
+    else
+        <:(E, T)
+    end
+end
+
+function eldeser(::Type{T}, ::Type{E}, key::Union{AbstractString,Symbol}, value_ptr::Ptr{YYJSONVal}) where {T,E}
+    return try
+        if yyjson_is_str(value_ptr) && check_types(E, AbstractString)
+            unsafe_string(yyjson_get_str(value_ptr))
```
Comment (AlexKlo) [inline src/De/DeJson.jl:256]: @gryumov 
This function check that expected type (field) corresponds to `value_ptr` type

File: src/De/DeJson.jl [hunk 10]
```diff
@@ -2,15 +2,359 @@ module DeJson
...
+                tuple_elements = Vector{Any}(undef, fieldcount(T))
+                for (i, type) in zip(eachindex(tuple_elements), fieldtypes(T))
+                    value_ptr = yyjson_arr_iter_next(iter_ptr)
+                    tuple_elements[i] = deser(type, value_ptr)
+                end
+            end
+            T(tuple_elements)
+        end
+    end
+end
```
Comment (AlexKlo) [inline src/De/DeJson.jl:203]: @gryumov 
In this function, the first condition is intended for pure Tuple, which in advance is unknown by the number of elements
The second branch for tuple {t1, t2, ...} or NTuple

File: src/De/DeJson.jl [hunk 11]
```diff
@@ -2,15 +2,359 @@ module DeJson
...
+        for (i, field_type) in zip(eachindex(type_elements), fieldtypes(T))
+            field_ptr = yyjson_arr_iter_next(iter_ptr)
+            value = field_ptr == C_NULL ? nulltype(field_type) : deser(T, field_type, field_ptr)
+            type_elements[i] = isempty(T, value) ? nulltype(field_type) : value
+        end
+        return T(type_elements...)
+    end
+end
+
+function typeof_yyjson_val(val_ptr::Ptr{YYJSONVal})
```
Comment (gryumov) [inline src/De/DeJson.jl:223]:
```suggestion
function typeof_yyjson_val(value_ptr::Ptr{YYJSONVal})
```

File: src/De/DeJson.jl [hunk 12]
```diff
@@ -2,15 +2,359 @@ module DeJson
...
+
+function deser(::PrimitiveType, ::Type{T}, value_ptr::Ptr{YYJSONVal}) where {T<:AbstractString}
+    return if yyjson_is_str(value_ptr)
+        unsafe_string(yyjson_get_str(value_ptr))
+    elseif yyjson_is_raw(value_ptr)
+        unsafe_string(yyjson_get_raw(value_ptr))
+    elseif yyjson_is_null(value_ptr)
+        nothing
+    else
+        throw(error("Expected a string for type `AbstractString`."))
```
Comment (gryumov) [inline src/De/DeJson.jl:55]:
```suggestion
        error("Expected a string for type `Symbol`.")
```

**Pull request merged**